### PR TITLE
WT-13389 Implement block_session extent unit tests

### PR DESCRIPTION
**Pull request by jiechenbo**
This tickets aims to test the block manager session part of the code. The idea behind the unit testing allows us to drill down into core functions of the block session code and check it's behaviour. There are a few technical changes I needed to make and would like to get reviews on:

- I configured the s_functions.py file to allow __wti functions to be tested outside of the module.
- I needed to move the WT_BLOCK_MGR_SESSION outside of the source file. I used #ifdef definitions to do this if HAVE_UNIT is enabled. 
- I created unit test functions version of all the internal functions used inside block_session.c
- I have added a block manager session creation inside the mock_session.c, that is used for almost all test cases.
- I have added a new unit test file that aims to test the core functions inside block_session.c

**Update**: This PR will only test block session ext functions.

### CODE REVIEW

**Comment by jiechenbo** [comment]
I will merge my PR after #10968 merges in

File: src/include/block.h
Comment (luke-pearson) [inline src/include/block.h:111]: This should sit in the unit test code if it is only used there.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:86]: In our C++ style declarations can happen where they are needed please the int  i = 0;.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:132]: You can declare and initialize on the same line in c++.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:103]: I think this is insufficient validation? The function does the following:

s- Callocs memory of size WT_EXT + skipdepth * 2 * pointer size. 
- Sets the ext->depth field which is whatever skipdepth was.

I can see we check that depth != 0 which makes sense but we don't check any of the other field in the memory that we now own. Technically the following should _also_ be 0:
- ext->off
- ext->size
- ext->next[0] to skipdepth should all be NULL pointers.

Now one could argue we're validating calloc but I think validating the skipdepth logic would be valuable.

File: src/include/block.h
Comment (luke-pearson) [reply src/include/block.h:111]: Oh right, it's used all through block_session.c, in that case just drop the #ifdef HAVE_UNITTEST and move it entirely IMO.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:113]: I think if we improve the validation functions, similar to what I have said above then this works better.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:57]: Another test is needed here, if `ext_cache_cnt` is 0 then `ext_cache` should be NULL.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:85]: This also needs an additional test that if `sz_cache_cnt` is 0 then `sz_cache` is NULL.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:117]: blocks -> block

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:123]: Why not 3 and not 1K? Or even 100K?

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:135]: How about `__ut_block_size_alloc(nullptr, nullptr)`?

Also this doesn't validate much of the returned structure, although again we could argue we're testing calloc more than the function itself.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:128]: In this test case we don't actually test what happens if you gave it zero after having existing items in the list, it should validate that the list doesn't grow unexpectedly. Similarly we've only gave it increasing allocations we should check that a decreasing value still allocates the correct number of blocks. Imagine for some reason it saves the max number it's been provided and only every allocates that many.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:139]: I think most of my __block_ext_prealloc comments apply here.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:165]: I would move these declarations below the first `SECTION` otherwise it is a bit confusing for the reader.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:22]: I would move these validation / cleanup functions as close to where they are used.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:203]: Here we should fill ext->next with junk and ensure it is cleared.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:245]: I'm guessing this needs discussion because it isn't testable? The pointer you pass it isn't cleared.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:260]: I would also be checking `ext_cache_cnt` here.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:284]: We could also check  session->block_manager_cleanup = __block_manager_session_cleanup here. It's also worth commenting why this is the case.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:292]: This doesn't validate that max is passed through correctly.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp:123]: IMO, i chose 3 to make sure that we are testing the cache. I don't really see much benefit from increasing from 3 other than the fact that we are stress testing the function. Let me know if you disagree and see benefit from increasing it.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp:135]: Unfortunately this will segfault, since it dereferences the sz pointer. Is it possible to test for a segfault?

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp:245]: Yes, we need more discussion on this. @marcbutler placed up an idea if possible to intercept the malloc and free functions and check if it has been incremented and decrement to test this out. 

It is also a matter of question if we want to even test these scenarios as well.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp:203]: I don't quite get what you mean here, the second allocation is meant to work as a cache mechanism. Therefore the ext->next[0] shouldn't be cleared in this place. what were you looking for? @luke-pearson

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: thanks fixed up now

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp:81]: I have expanded the validation to check for this now.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp:91]: Done.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: I have added that check now thanks.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: done.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: thanks

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: Makes sense, thank you for those test cases, I have added them now.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: Yep, fixed most of them.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp:125]: Done

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: I have organised the ext functions and size functions together along with the validation, cleanup functions. Let me know if it worked better.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp:220]: Done!

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp:258]: I could only check if it is not null, the __block_manager_session_cleanup function is internal to the file. We can't expose it as well because that would create a different address for the function e.g. __ut_block_manager_session_cleanup has different address to __block_manager_session_cleanup

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: I have added test cases for this.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jeremythorp) [inline test/unittest/tests/block/test_block_session.cpp:233]: I suggest moving the allocation of `ext`, `ext2` and `ext3` to just above the line where `ext` is first used.

File: src/include/block.h
Comment (luke-pearson) [reply src/include/block.h:111]: @jiechenbo can we action this one? It'll cleanup the PR a bit.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [reply test/unittest/tests/block/test_block_session.cpp:204]: I think the best thing to do here would be to update the function definition of `__wti_block_ext_free` so that it takes a ** pointer and therefore can have the original pointer NULL'd out, which is what our __wt_free does anyway at this stage. 

cc: @jeremythorp

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [reply test/unittest/tests/block/test_block_session.cpp:258]: That's true, then verifying it was NULL and now isn't NULL would be valuable.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:327]: This file is too big to have an effective review, it may only be test code but already numerous issues have been called out. To have an effective discussion I think all of the functions being tested from here down should be moved to a subsequent ticket.

File: src/include/block.h
Comment (jiechenbo) [reply src/include/block.h:110]: Fixed up now

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: thanks fixed it up.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: I agree, i have split the ticket into multiple different tickets. This ticket will focus on the size functions only.

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:43]: For each WT_EXT you walk you should call `validate_ext_block` on it. If you split functions out a bit more you can do this with 6 functions:

1. validate_ext_block
2. free_ext_block
3. validate_and_free_ext_block
4. validate_ext_list
5. free_ext_list
6. validate_and_free_ext_list

This would look like this:
```
void
free_ext_block(WT_EXT *ext) {
    __wt_free(nullptr, ext);
}

void
validate_ext_block(WT_EXT *ext)
{
    REQUIRE(ext != nullptr);
    REQUIRE(ext->depth != 0);
    REQUIRE(ext->size == 0);
    REQUIRE(ext->off == 0);

    for (int i = 0; i < ext->depth; i++) {
        REQUIRE(ext->next[i + ext->depth] == nullptr);
    }
}

void
validate_and_free_ext_block(WT_EXT *ext) {
    validate_ext_block(ext);
    free_ext_block(ext);
}

void
free_ext_list(WT_BLOCK_MGR_SESSION *bms)
{
    WT_EXT *curr = bms->ext_cache;
    while (curr != nullptr) {
        WT_EXT *tmp = curr;
        curr = curr->next[0];
        free_ext_block(tmp);
    }
}

void
validate_ext_list(WT_BLOCK_MGR_SESSION *bms, int expected_items)
{
    REQUIRE(bms != nullptr);
    WT_EXT *curr = bms->ext_cache;

    if (bms->ext_cache_cnt == 0)
        REQUIRE(bms->ext_cache == nullptr);

    REQUIRE(bms->ext_cache_cnt == expected_items);
    for (int i = 0; i < expected_items; i++) {
        validate_ext_block(curr);
        REQUIRE(bms->ext_cache != nullptr);
        curr = curr->next[0];
    }
    REQUIRE(curr == nullptr);
}
```

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [reply test/unittest/tests/block/test_block_session.cpp:162]: I'm referring to testing this code branch: 
```
        /* Clear any left-over references. */
        for (i = 0; i < ext->depth; ++i)
            ext->next[i] = ext->next[i + ext->depth] = NULL;
```

I think we could test that?

File: test/unittest/tests/block/test_block_session.cpp
Comment (luke-pearson) [inline test/unittest/tests/block/test_block_session.cpp:252]: At this stage we've discarded everything in the extent list. Shouldn't we put items back in?

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: Makes sense, i have done that now thanks!

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp:261]: No, for each unit test section, the test case gets setup afresh again. So the extent list is populated again at this.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp:174]: Oh yes, i see what you mean, I have added junk next now.

File: dist/function.py
Comment (luke-pearson) [reply dist/function.py:319]: @jiechenbo could you take a look at Tod's solution? This PR is closer to getting merged so I think we'd want to bring his solution across and then merge this PR.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: I have created a ticket and FIXME comment [WT-13451](https://jira.mongodb.org/browse/WT-13451)

File: dist/function.py
Comment (jiechenbo) [reply dist/function.py:323]: Add the functions.py from the tods changes. I did modify it a bit, let me know what you think.

Comment (jeremythorp) [reply dist/function.py]: Thanks @tod-johnson-mongodb for picking that up - your PR has an important improvement there.
Thanks @jiechenbo for copying over @tod-johnson-mongodb 's improvement, which you've then slightly simplified.
Great collaboration from you both - thanks!

File: test/unittest/tests/block/test_block_session.cpp
Comment (jeremythorp) [inline test/unittest/tests/block/test_block_session.cpp:44]: As this code is C++, and our coding standard allows it, variables can be declared for the smaller scope. I suggest moving the declaration of `curr` to when it is first required - ie immediately before the `for` loop.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jeremythorp) [inline test/unittest/tests/block/test_block_session.cpp:84]: As above, the declaration of `ext` can be moved to when it is first required.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jeremythorp) [inline test/unittest/tests/block/test_block_session.cpp:151]: `cached_ext` can be declared just before it is used.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jeremythorp) [inline test/unittest/tests/block/test_block_session.cpp:186]: cached_ext can be declared just before it is used.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jeremythorp) [inline test/unittest/tests/block/test_block_session.cpp:227]: `ext2` can be declared just before it is used.

File: test/unittest/tests/wrappers/mock_session.cpp
Comment (jeremythorp) [inline test/unittest/tests/wrappers/mock_session.cpp:52]: We should rename `throwIfNonZero` to the current naming convention, however I suggest we do that in a seperate ticket.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: Thanks fixed. [5f71e2e](https://github.com/wiredtiger/wiredtiger/pull/10963/commits/5f71e2efe324cb7f0f225d03d07a9041839c75f4)

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: [5f71e2e](https://github.com/wiredtiger/wiredtiger/pull/10963/commits/5f71e2efe324cb7f0f225d03d07a9041839c75f4)

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: [5f71e2e](https://github.com/wiredtiger/wiredtiger/pull/10963/commits/5f71e2efe324cb7f0f225d03d07a9041839c75f4)

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: [5f71e2e](https://github.com/wiredtiger/wiredtiger/pull/10963/commits/5f71e2efe324cb7f0f225d03d07a9041839c75f4)

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: [5f71e2e](https://github.com/wiredtiger/wiredtiger/pull/10963/commits/5f71e2efe324cb7f0f225d03d07a9041839c75f4)

File: test/unittest/tests/wrappers/mock_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/wrappers/mock_session.cpp:52]: Hey jeremy, i agree, [WT-13405](https://jira.mongodb.org/browse/WT-13405) will be covering the change.

Comment (jeremythorp) [reply test/unittest/tests/wrappers/mock_session.cpp:52]: Perfect - thanks Jie.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jeremythorp) [reply test/unittest/tests/block/test_block_session.cpp]: That was quick :-)

File: test/unittest/tests/block/test_block_session.cpp
Comment (jeremythorp) [inline test/unittest/tests/block/test_block_session.cpp]: The line of code that declares `ext` and `ext2` is not quite correct.
As written it declares `ext` as a `WT_EXT*` and `ext2` as a `WT_EXT` (as an object not a pointer). 
It should be
`        WT_EXT *ext, *ext2;`
or 
```
        WT_EXT *ext;
        WT_EXT *ext2;
```

Comment (jeremythorp) [reply test/unittest/tests/block/test_block_session.cpp]: Now fixed.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jeremythorp) [inline test/unittest/tests/block/test_block_session.cpp:270]: It took me a moment to match the section name ("Discard nothing in the extent list") with the parameter 3 passed to both `__ut_block_ext_discard()` and `validate_and_free_ext_list()`. Once I realised that the exit list had 3 items to start with, then that became clear.
I suggest either updating the section name, or adding a comment, to clarify that the 'discard nothing' is because the the number of items is already correct.

**Review by luke-pearson — APPROVED** [review]
LGTM.

File: test/unittest/tests/block/test_block_session.cpp
Comment (jiechenbo) [reply test/unittest/tests/block/test_block_session.cpp]: Updated the section description [37eb6eb](https://github.com/wiredtiger/wiredtiger/pull/10963/commits/37eb6eb87910c7eae95b57d4d6bccdae2ca69fba)

Comment (jeremythorp) [reply test/unittest/tests/block/test_block_session.cpp]: Thanks!

**Review by jeremythorp — APPROVED** [review]
LGTM!

**Pull request merged**
